projects/npm-tools/packages/npm-scripts/test/typescript/getTypeScriptBuildOrder.js

```diff
@@ -0,0 +1,234 @@
+		expect(getTypeScriptBuildOrder({a, c, d, b})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({a, d, b, c})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({a, d, c, b})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, a, c, d})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, a, d, c})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, c, a, d})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, c, d, a})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, d, a, c})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({b, d, c, a})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, a, b, d})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, a, d, b})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, b, d, a})).toEqual([c, a, d, b]);
+		expect(getTypeScriptBuildOrder({c, b, d, a})).toEqual([c, a, d, b]);
+
+		// But note that we can get different (but still valid) orderings:
+
+		expect(getTypeScriptBuildOrder({c, d, a, b})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({c, d, b, a})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, a, b, c})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, a, c, b})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, b, a, c})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, b, c, a})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, c, a, b})).toEqual([c, d, a, b]);
+		expect(getTypeScriptBuildOrder({d, c, b, a})).toEqual([c, d, a, b]);
+
+		// And we can get a similar effect by changing the order of
+		// dependencies _inside_ "b"; again, a different (but still
+		// valid) ordering.
+
+		const b2 = {name: 'b2', dependencies: {d: '*', a: '*'}};
+
+		expect(getTypeScriptBuildOrder({b2, a, c, d})).toEqual([c, d, a, b2]);
+
+		/* eslint-enable sort-keys */
+	});
+
+	it('handles a more complex example', () => {
+
+		//            Y
+		//            | \
+		//        Z   X  \   ^
+		//       /|\ /|   \  | dependencies point upwards
+		//      J K L M N |  |
+		//      | | |\|/ /
+		//      A B C D-/
+
+		/* eslint-disable sort-keys */
+
+		const a = {name: 'a', dependencies: {j: '*'}};
+		const b = {name: 'b', dependencies: {k: '*'}};
+		const c = {name: 'c', dependencies: {l: '*'}};
+		const d = {name: 'd', dependencies: {l: '*', m: '*', n: '*', y: '*'}};
+		const j = {name: 'j', dependencies: {z: '*'}};
+		const k = {name: 'k', dependencies: {z: '*'}};
+		const l = {name: 'l', dependencies: {z: '*', x: '*'}};
+		const m = {name: 'm', dependencies: {x: '*'}};
+		const n = {name: 'n', dependencies: {}};
+		const x = {name: 'x', dependencies: {y: '*'}};
+		const y = {name: 'y', dependencies: {}};
+		const z = {name: 'z', dependencies: {}};
+
+		expect(
+			getTypeScriptBuildOrder({a, b, c, d, j, k, l, m, n, x, y, z})
+		).toEqual([z, j, a, k, b, y, x, l, c, m, n, d]);
```

Comment:
Out of curiosity... where would an unrelated dependency `U` (or a tree) that has no shared `dependencies` with the rest of the tree (A,B,C...X,Y,Z) sit in the end result? Before, after, in between? Is it reproducible?

In this case, I'm wondering if loading unrelated deps first or sorting by smaller graphs first could result in freeing up work faster. Of course, this hardly applies here so I'm thinking more about a JS loading context, so feel free to ignore this and everything else I said 😂

Reply:
> Before, after, in between? It depends?

Without having tested it, I would say it depends on where you list it in the graph. JS will iterate over objects in insertion order, so the possible orderings are:

- If `U` is inserted first, it will appear first in the output.
- If `U` is inserted last, it will appear last in the output.
- If `U` is inserted somewhere in the middle, it will appear after all the nodes reachable via DFS from the preceding nodes have been visited.

> In this case, I'm wondering if loading unrelated deps first or sorting by smaller graphs first could result in freeing up work faster. 

An excellent thought to bear in mind when we start scaling this. Imagine we have 200 TS projects in the future. We can break the graph up into connected components and run the builds for those in parallel. Of course, one would expect a very large part of the graph to be one huge connected component (many things, for example, will depend on `frontend-js-web` and/or `frontend-js-react-web` and/or `frontend-js-state-web`), but we can look at parallelizing the build within connected components as well. But obviously, no sense in doing that until we actually need to.

Reply:
Sounds like an interesting problem space to explore in the future!!

![yum](https://user-images.githubusercontent.com/905006/113157583-f0266400-9232-11eb-9948-b6deb637ae4c.gif)